mlflow/models/signature.py

```diff
@@ -0,0 +1,449 @@
+    raise MlflowException("Unsupported numpy data type '{0}', kind '{1}'".format(
+        col.dtype, col.dtype.kind))
+
+
+def _map_spark_type(x) -> DataType:
+    import pyspark.sql.types
+    if isinstance(x, pyspark.sql.types.NumericType):
+        if isinstance(x, pyspark.sql.types.IntegralType):
+            if isinstance(x, pyspark.sql.types.LongType):
+                return DataType.long
+            else:
+                return DataType.integer
+        elif isinstance(x, pyspark.sql.types.FloatType):
+            return DataType.float
+        elif isinstance(x, pyspark.sql.types.DoubleType):
+            return DataType.double
+    elif isinstance(x, pyspark.sql.types.BooleanType):
+        return DataType.boolean
+    elif isinstance(x, pyspark.sql.types.StringType):
+        return DataType.string
+    elif isinstance(x, pyspark.sql.types.BinaryType):
+        return DataType.binary
+    else:
+        raise Exception("Unsupported Spark Type '{}', MLflow schema is only supported for scalar "
+                        "Spark types.".format(type(x)))
+
+
+def _is_spark_df(x) -> bool:
+    try:
+        return isinstance(x, importlib.import_module("pyspark.sql.dataframe").DataFrame)
+    except ImportError:
+        return False
+
+
+def _infer_schema(data: MlflowModelDataset) -> Schema:
+    if hasattr(data, "__len__") and len(data) == 0:
+        return Schema([])
+    if isinstance(data, dict):
+        res = []
+        for col in data.keys():
+            ary = data[col]
+            if not isinstance(ary, np.ndarray):
+                raise TypeError("Data in the dictionary must be of type numpy.ndarray")
+            dims = len(ary.shape)
+            if dims == 1:
+                res.append(ColSpec(col, _map_numpy_dtype(ary)))
+            else:
+                raise MlflowException("Data in the dictionary must be 1-dimensional, "
+                                      "got shape {}".format(ary.shape))
+        return Schema(res)
+
+    if isinstance(data, pd.DataFrame):
+        return Schema([ColSpec(col, _map_numpy_dtype(data[col].values)) for col in data.columns])
+    elif isinstance(data, np.ndarray):
+        if len(data.shape) > 2:
+            raise MlflowException("Multidimensional arrays (aka tensors) are not supported, "
+                                  "got array with shape {}".format(data.shape))
+        if data.dtype == np.dtype("O"):
+            df = pd.DataFrame(data).infer_objects()
+            schema = _infer_schema(df)
+            return Schema([ColSpec(None, t) for t in schema.column_types()])
+        else:
+            if len(data.shape) == 2:
+                array_type = _map_numpy_dtype(data[0])
+                return Schema([ColSpec(None, array_type)] * data.shape[1])
+            else:
+                array_type = _map_numpy_dtype(data)
+                return Schema([ColSpec(None, array_type)])
+    elif _is_spark_df(data):
```

Comment:
Is this possible given the type signature?

Reply:
yes